pkg/controller/crdmirroring/crdhandler/clustergroup.go

```diff
@@ -0,0 +1,163 @@
+
+package crdhandler
+
+import (
+	"context"
+	"reflect"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	crd "github.com/vmware-tanzu/antrea/pkg/apis/crd/v1alpha2"
+	crdclientset "github.com/vmware-tanzu/antrea/pkg/client/clientset/versioned"
+	crdlister "github.com/vmware-tanzu/antrea/pkg/client/listers/crd/v1alpha2"
+	"github.com/vmware-tanzu/antrea/pkg/controller/crdmirroring/types"
+	legacycore "github.com/vmware-tanzu/antrea/pkg/legacyapis/core/v1alpha2"
+	legacycrdclientset "github.com/vmware-tanzu/antrea/pkg/legacyclient/clientset/versioned"
+	legacycorelister "github.com/vmware-tanzu/antrea/pkg/legacyclient/listers/core/v1alpha2"
+)
+
+type ClusterGroupHandler struct {
+	lister       crdlister.ClusterGroupLister
+	legacyLister legacycorelister.ClusterGroupLister
+	client       crdclientset.Interface
+	legacyClient legacycrdclientset.Interface
+}
+
+func NewClusterGroupHandler(lister crdlister.ClusterGroupLister,
+	legacyLister legacycorelister.ClusterGroupLister,
+	client crdclientset.Interface,
+	legacyClient legacycrdclientset.Interface) types.MirroringHandler {
+	mc := &ClusterGroupHandler{
+		lister:       lister,
+		legacyLister: legacyLister,
+		client:       client,
+		legacyClient: legacyClient,
+	}
+	return mc
+}
+
+// GetNewObject gets the mirroring new ClusterGroup struct
+func (c *ClusterGroupHandler) GetNewObject(namespace, name string) (metav1.Object, error) {
+	lister := c.lister
+	return lister.Get(name)
+}
+
+// AddNewObject creates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) AddNewObject(obj metav1.Object) error {
+	l := obj.(*legacycore.ClusterGroup)
+	n := c.buildNewObject(l)
+	client := c.client.CrdV1alpha2().ClusterGroups()
+	_, err := client.Create(context.TODO(), n, metav1.CreateOptions{})
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// SyncObject updates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) SyncObject(legacyObj, newObj metav1.Object) error {
+	if !c.deepEqualSpecAndLabels(legacyObj, newObj) {
+		n := c.syncNewObject(legacyObj, newObj)
+		newClient := c.client.CrdV1alpha2().ClusterGroups()
+		_, err := newClient.Update(context.TODO(), n, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+
+	if !c.deepEqualStatus(legacyObj, newObj) {
+		l := c.syncLegacyObject(legacyObj, newObj)
+		legacyClient := c.legacyClient.CoreV1alpha2().ClusterGroups()
+		_, err := legacyClient.UpdateStatus(context.TODO(), l, metav1.UpdateOptions{})
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+//DeleteNewObject deletes the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) DeleteNewObject(namespace, name string) error {
+	client := c.client.CrdV1alpha2().ClusterGroups()
+	return client.Delete(context.TODO(), name, metav1.DeleteOptions{})
+}
+
+// StaleLegacyObject updates the mirroring new ClusterGroup
+func (c *ClusterGroupHandler) StaleLegacyObject(legacyObj, newObj metav1.Object) error {
+	delete(newObj.GetAnnotations(), types.ManagedBy)
```

Comment:
objects returned by Lister is supposed to be read-only, otherwise race condition might happen or index may be messed up. Create a deepcopy before updating an object.

Reply:
Got it.

Reply:
Got it.